src/listpack.c

```diff
@@ -944,6 +952,25 @@ unsigned char *lpInsertInteger(unsigned char *lp, long long lval, unsigned char
     return lpInsert(lp, NULL, intenc, enclen, p, where, newp);
 }
 
+/* This is just a wrapper for lpInsert() to directly take a double.
+ * If the double happens have a value representable by long long, it'll be inserted as an integer,
+ * if not, it'll be converted to string.
+ * The advantage of exposing such an interface is that the caller doesn't convert doubles to strings,
+ * and then the listpack code tries to see if they happen to be integers (inefficient). */
+unsigned char *lpInsertDouble(unsigned char *lp, double dval, unsigned char *p, int where, unsigned char **newp) {
+    long long lval;
+    if (double2ll(dval, &lval)) {
+        uint64_t enclen; /* The length of the encoded element. */
+        unsigned char intenc[LP_MAX_INT_ENCODING_LEN];
+        lpEncodeIntegerGetType(lval, intenc, &enclen);
+        return lpInsert(lp, NULL, intenc, enclen, p, where, newp);
+    } else {
+        char sval[MAX_D2STRING_CHARS];
+        int slen = d2string(sval,sizeof(sval),dval);
+        return lpInsert(lp, (unsigned char *)sval, NOT_INT, slen, p, where, newp);
```

Comment:
Will this solve the problem?

```suggestion
        int is_huge = dval < -4503599627370495 || dval > 4503599627370495;
        return lpInsert(lp, (unsigned char *)sval, is_huge ? NULL : NOT_INT, slen, p, where, newp);
```